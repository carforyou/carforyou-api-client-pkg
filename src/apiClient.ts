export interface ApiClientConfig {
  carServiceUrl?: string
  searchServiceUrl?: string
  catalogueServiceUrl?: string
  dealerServiceUrl?: string
  optionServiceUrl?: string
  analyticsServiceUrl?: string
  userNotificationServiceUrl?: string
  tokenRefreshServiceUrl?: string
  debug?: boolean
}

export interface Tokens {
  accessToken?: string
  refreshToken?: string
}

export interface Handlers {
  onAccessTokenUpdate?: (token: string) => void
  onFailedTokenRefresh?: () => void
}

class ApiClient {
  static instance: ApiClient

  configuration: ApiClientConfig = {}
  tokens: Tokens = {}
  handlers: Handlers = {}
  version = "v1"

  constructor() {
    ApiClient.instance = this
  }

  public configure(configuration: ApiClientConfig): void {
    if (Object.keys(this.configuration).length) {
      throw new Error("Owerwriting API client configuration")
    }

    Object.keys(configuration).forEach((key) => {
      this.configuration[key] = configuration[key]
    })
  }

  public setTokens(tokens: Tokens): void {
    this.tokens.accessToken = tokens.accessToken
    this.tokens.refreshToken = tokens.refreshToken
  }

  public setHandlers(handlers: Handlers): void {
    this.handlers.onAccessTokenUpdate = handlers.onAccessTokenUpdate
    this.handlers.onFailedTokenRefresh = handlers.onFailedTokenRefresh
  }
}

const instance = new ApiClient()
Object.freeze(instance)

export default instance
