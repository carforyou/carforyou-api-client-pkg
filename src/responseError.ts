export class ResponseError extends Error {
  code?: string
  status: number
  statusCode: number
  response: Response

  constructor(response: Response) {
    super()
    this.name = "ResponseError"
    this.message = "API ResponseError: " + response.statusText
    this.response = response

    this.code = response.status === 404 ? "ENOENT" : null
    this.status = this.statusCode = response.status || 500
  }
}
