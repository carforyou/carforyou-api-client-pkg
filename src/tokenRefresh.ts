import apiClient from "./apiClient"
import { Service, postData } from "./base"

const refreshToken = async () => {
  const { refresh_token, access_token } = await postData(
    Service.TOKEN_REFRESH,
    "refresh-token",
    { token: apiClient.tokens.refreshToken }
  )

  apiClient.setTokens({
    accessToken: access_token,
    refreshToken: refresh_token,
  })
  if (apiClient.handlers.onAccessTokenUpdate) {
    apiClient.handlers.onAccessTokenUpdate(apiClient.tokens.accessToken)
  }
}

const refreshTokenAndRetry = async (apiCall) => {
  try {
    await refreshToken()
    return apiCall()
  } catch (error) {
    if (error.name === "ResponseError") {
      const status = error.status.toString()
      if (
        (status.startsWith("4") || status.startsWith("5")) &&
        apiClient.handlers.onFailedTokenRefresh
      ) {
        apiClient.handlers.onFailedTokenRefresh()
      }

      return
    }

    throw error
  }
}

export const withTokenRefresh = async (apiCall) => {
  try {
    return await apiCall()
  } catch (error) {
    if (error.name === "ResponseError" && error.status.toString() === "401") {
      return refreshTokenAndRetry(apiCall)
    }

    throw error
  }
}
