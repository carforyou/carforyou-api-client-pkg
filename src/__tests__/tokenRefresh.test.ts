import apiClient from "../apiClient"
import { withTokenRefresh } from "../tokenRefresh"
import { ResponseError } from "../responseError"

describe("apiClient", () => {
  describe("#withTokenRefresh", () => {
    const apiCall = jest.fn(() => Promise.resolve({ ok: true }))

    afterEach(() => {
      apiCall.mockClear()
      fetchMock.resetMocks()
    })

    describe("on a failed api call that is unauthorized", () => {
      const mockRefresh = jest.fn()
      const mockRefreshFailure = jest.fn()

      beforeEach(() => {
        apiClient.setHandlers({
          onFailedTokenRefresh: mockRefreshFailure,
          onAccessTokenUpdate: mockRefresh
        })

        apiCall.mockImplementationOnce(() => {
          throw new ResponseError({ status: 401 })
        })
      })

      describe("and a successfull refresh", () => {
        beforeEach(() => {
          fetchMock.mockResponseOnce(
            JSON.stringify({
              refresh_token: "new refresh",
              access_token: "new access"
            })
          )
        })

        it("retries the API call", async () => {
          const result = await withTokenRefresh(apiCall)

          expect(apiCall).toHaveBeenCalledTimes(2)
          expect(result).toEqual({ ok: true })
        })

        it("updates the tokens", async () => {
          await withTokenRefresh(apiCall)

          expect(apiClient.tokens.accessToken).toEqual("new access")
          expect(apiClient.tokens.refreshToken).toEqual("new refresh")
        })

        it("calls the update handler with new access token", async () => {
          await withTokenRefresh(apiCall)

          expect(mockRefresh).toHaveBeenCalledWith("new access")
        })
      })

      describe("and a failed refresh", () => {
        beforeEach(() => {
          fetchMock.mockResponseOnce(JSON.stringify({ ok: false }), {
            status: 400
          })
        })

        it("calls refresh failure", async () => {
          await withTokenRefresh(apiCall)

          expect(mockRefreshFailure).toHaveBeenCalled()
        })
      })
    })

    describe("on a failed api call that's not unauthorized", () => {
      beforeEach(() => {
        apiCall.mockImplementationOnce(() => {
          throw new ResponseError({ status: 400 })
        })
      })

      it("only tries the api call once", async () => {
        try {
          await withTokenRefresh(apiCall)
        } catch {
          //
        }

        expect(apiCall).toHaveBeenCalledTimes(1)
      })

      it("does not try to refresh the token", async () => {
        try {
          await withTokenRefresh(apiCall)
        } catch {
          //
        }

        expect(fetchMock).toHaveBeenCalledTimes(0)
      })

      it("rethrows the error", async () => {
        await expect(
          withTokenRefresh(apiCall)
        ).rejects.toThrowErrorMatchingSnapshot()
      })
    })
  })
})
