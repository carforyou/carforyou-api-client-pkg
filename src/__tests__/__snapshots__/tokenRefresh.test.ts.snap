// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apiClient #withTokenRefresh on a failed api call that's not unauthorized rethrows the error 1`] = `"API ResponseError: undefined"`;
